src/HotChocolate/CostAnalysis/src/CostAnalysis/CostAnalysisMiddleware.cs

```diff
@@ -0,0 +1,108 @@
+using HotChocolate.CostAnalysis.Caching;
+using HotChocolate.Execution;
+using HotChocolate.Execution.Processing;
+using HotChocolate.Language;
+using HotChocolate.Validation;
+using Microsoft.Extensions.DependencyInjection;
+using static HotChocolate.CostAnalysis.WellKnownContextData;
+
+namespace HotChocolate.CostAnalysis;
+
+internal sealed class CostAnalysisMiddleware(
+    RequestDelegate next,
+    CostAnalysisOptions options,
+    DocumentValidatorContextPool contextPool,
+    ICostMetricsCache cache,
+    CostAnalysisVisitor costAnalysisVisitor)
+{
+    public async ValueTask InvokeAsync(IRequestContext context)
+    {
+        if (context.Document is not null && context.OperationId is not null)
+        {
+            var document = context.Document;
+            var operationDefinition =
+                context.Operation?.Definition ??
+                document.GetOperation(context.Request.OperationName);
+
+            var validatorContext = contextPool.Get();
+
+            try
+            {
+                if (!cache.TryGetCostMetrics(context.OperationId, out var costMetrics))
+                {
+                    PrepareContext(context, document, validatorContext);
+
+                    costAnalysisVisitor.Visit(operationDefinition, validatorContext);
+
+                    costMetrics = (CostMetrics)validatorContext.ContextData[RequestCosts]!;
+
+                    cache.TryAddCostMetrics(context.OperationId, costMetrics);
+                }
```

Comment:
@PascalSenn how shall we report the costs to metrics? Just like everything else with the diagnostic events?